frame/staking/src/offchain_election.rs

```diff
@@ -155,19 +163,192 @@ pub fn get_balancing_iters<T: Trait>() -> usize {
 	}
 }
 
+/// Find the maximum `len` that a compact can have in order to fit into the block weight.
+///
+/// This only returns a value between zero and `size.nominators`.
+pub fn maximum_compact_len<W: crate::WeightInfo>(
+	winners_len: u32,
+	size: ElectionSize,
+	max_weight: Weight,
+) -> u32 {
+	use sp_std::cmp::Ordering;
+
+	if size.nominators < 1 {
+		return size.nominators;
+	}
+
+	let max_voters = size.nominators.max(1);
+	let mut voters = max_voters;
+
+	// helper closures.
+	let weight_with = |voters: u32| -> Weight {
+		W::submit_solution_better(
+			size.validators.into(),
+			size.nominators.into(),
+			voters,
+			winners_len,
+		)
+	};
+
+	let next_voters = |current_weight: Weight, voters: u32, step: u32| -> Result<u32, ()> {
+		match current_weight.cmp(&max_weight) {
+			Ordering::Less => {
+				let next_voters = voters.checked_add(step);
+				match next_voters {
+					Some(voters) if voters < max_voters => Ok(voters),
+					_ => Err(()),
+				}
+			},
+			Ordering::Greater => voters.checked_sub(step).ok_or(()),
+			Ordering::Equal => Ok(voters),
+		}
+	};
+
+	// First binary-search the right amount of voters
+	let mut step = voters / 2;
+	let mut current_weight = weight_with(voters);
+	while step > 0 {
+		match next_voters(current_weight, voters, step) {
+			// proceed with the binary search
+			Ok(next) if next != voters => {
+				voters = next;
+			},
+			// we are out of bounds, break out of the loop.
+			Err(()) => {
+				break;
+			},
+			// we found the right value - early exit the function.
+			Ok(next) => return next
+		}
+		step = step / 2;
+		current_weight = weight_with(voters);
+	}
+
+
+	// Time to finish.
+	// We might have reduced less than expected due to rounding error. Reduce linearly one last
+	// time.
+	let mut initialized = false;
+	let mut prev_voters = voters;
+	loop {
+		let next = next_voters(current_weight, voters, 1).unwrap_or(voters);
+		// no change? break.
+		if next == voters {
+			break
+		}
+
+		// then we are stuck in a loop;
+		if prev_voters == next {
+			break;
+		}
+
+		// skip 1 loop, then update these 3 variables in a sequence:
+		// prev_voters --> voter --> next
+		if !initialized {
+			initialized = true
+		} else {
+			prev_voters = voters
+		}
+
+		voters = next;
+		current_weight = weight_with(voters);
+	}
```

Comment:
this loop doesn't check if voters is beyond the limit or not. I added `assert!(weight_with(voters) < max_weight);` and it fails the tests

Reply:
> EDIT: maybe this loop could be written by just adding 1 until we are beyond the maximum weight and then removing 1 until we are below the maxium weight.

This is exactly how it was before,  I think both are fine. 

The assertion is fine if you allow `<=`. In my test cases I have:

```
assert_eq!(maximum_compact_len::<Staking>(0, size, 1000), 1);
```

Which means actually the intended logic is to allow `<=`.

Reply:
ah my bad indeed test should be `<=` but still the function can result in `weight_with(voters) > max_weight` in this test: 
```
		let size = ElectionSize {
			validators: 0,
			nominators: 10,
		};
assert_eq!(maximum_compact_len::<Staking>(0, size, 2001), 3);
```